parlai/scripts/interactive_web.py

```diff
@@ -0,0 +1,244 @@
+#!/usr/bin/env python3
+
+# Copyright (c) Facebook, Inc. and its affiliates.
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+"""
+Basic script which allows human keyboard input to talk to a trained
+model on a web server.
+"""
+
+
+from http.server import BaseHTTPRequestHandler, HTTPServer
+from parlai.scripts.interactive import setup_args
+from parlai.core.agents import create_agent
+from parlai.core.worlds import create_task
+from parlai.core.params import ParlaiParser
+
+import json
+import time
+
+HOST_NAME = 'localhost'
+PORT_NUMBER = 8080
+SHARED = {}
+STYLE_SHEET = "https://cdnjs.cloudflare.com/ajax/libs/bulma/0.7.4/css/bulma.css"
+FONT_AWESOME = "https://use.fontawesome.com/releases/v5.3.1/js/all.js"
+WEB_HTML = """
+<html>
+    <link rel="stylesheet" href={} />
+    <script defer src={}></script>
+    <head><title> Interactive Run </title></head>
+    <body>
+        <div class="columns">
+            <div class="column is-three-fifths is-offset-one-fifth">
+              <section class="hero is-info is-large has-background-light has-text-grey-dark	">
+                <div id="parent" class="hero-body">
+                    <article class="media">
+                      <figure class="media-left">
+                        <span class="icon is-large">
+                          <i class="fas fa-robot fas fa-2x"></i>
+                        </span>
+                      </figure>
+                      <div class="media-content">
+                        <div class="content">
+                          <p>
+                            <strong>Model</strong>
+                            <br>
+                            Enter a message, and the model will respond interactively.
+                          </p>
+                        </div>
+                      </div>
+                    </article>
+                </div>
+                <div class="hero-foot column is-three-fifths is-offset-one-fifth">
+                  <form id = "interact">
+                      <div class="field is-grouped">
+                        <p class="control is-expanded">
+                          <input class="input" type="text" id="userIn" placeholder="Type in a message">
+                        </p>
+                        <p class="control">
+                          <button id="respond" type="submit" class="button has-text-white-ter has-background-grey-dark">
+                            Submit
+                          </button>
+                        </p>
+                      </div>
+                  </form>
+                </div>
+              </section>
+            </div>
+        </div>
+
+        <script>
+            function createChatRow(agent, text) {{
+
+            var article = document.createElement("article");
```

Comment:
indentation is off.

Note: Talked to Jai offline and we discussed cleaning up the code a bit by introducing this helper method to generate chat rows. Additionally, instead of introducing another dependency, we are trying to use a lightweight approach to templating here. There are some downsides though, e.g. curly braces have to be escaped, e.g. ``{{`` and ``}}`` and also we have to be more mindful of HTML injection. In this case, `createTextNode` protects us in the case of the text. The agent name could technically be a vector for injection, however it is not modifiable by the user.